proposal.rst

```diff
@@ -369,6 +369,9 @@ Plan :
         with exceptions, they are panic defer and recover.A Panic is similar to an exception which can occur an runtime exception.
         C++ exceptions can be handled in go from swig wrappers as follows(https://github.com/srinivasyadav18/xapian-gsoc-plan/blob/master/example.i#L16).
         Which ever class function throws an exception in c++ , the wrapped function in Go returns the error as value.
+        Database error handling - https://github.com/srinivasyadav18/xapian-gsoc-plan/blob/master/go.i#L77
```

Comment:
We really need to find a way to automate such error handling - having to write this by hand for every wrapped method will make the Go bindings too hard to maintain going forwards.

For a typical new C++ method, we get automatic wrapping for all the bindings for **zero** extra work.  In some cases, extra work to get a new C++ method wrapped is needed, but it's rare.

Having to do something for every new wrapped method scales badly in terms of developer effort, and we definitely need to avoid that.  We've had non-SWIG-based bindings which required manual update before (XS bindings for Perl and JNI bindings for Java) and they constantly lagged behind the C++ API, and that frustrated both users and developers.

Reply:
> We really need to find a way to automate such error handling - having to write this by hand for every wrapped method will make the Go bindings too hard to maintain going forwards.
> 
> For a typical new C++ method, we get automatic wrapping for all the bindings for **zero** extra work. In some cases, extra work to get a new C++ method wrapped is needed, but it's rare.
> 
> Having to do something for every new wrapped method scales badly in terms of developer effort, and we definitely need to avoid that. We've had non-SWIG-based bindings which required manual update before (XS bindings for Perl and JNI bindings for Java) and they constantly lagged behind the C++ API, and that frustrated both users and developers.

Since errors are treated as values in Golang I am explicitly hand wrapping for the function that throws exception. I would work on it and find a way to automate for wrapping error  in community bonding period.

Reply:
> We really need to find a way to automate such error handling - having to write this by hand for every wrapped method will make the Go bindings too hard to maintain going forwards.
> 
> For a typical new C++ method, we get automatic wrapping for all the bindings for **zero** extra work. In some cases, extra work to get a new C++ method wrapped is needed, but it's rare.
> 
> Having to do something for every new wrapped method scales badly in terms of developer effort, and we definitely need to avoid that. We've had non-SWIG-based bindings which required manual update before (XS bindings for Perl and JNI bindings for Java) and they constantly lagged behind the C++ API, and that frustrated both users and developers.

Hey do you mean to say if a new c++ method gets added then that should get rewrapped automatically instead of hand wrapping?if that's the case..I think most of the work is in writing a script which generates the wrappers automatically. You are correct, we get automatic wrapping for other bindings with no extra work, but the swig support for Go doesn't produce the API which a user can use. So this means that we need to provide second level of wrapper (kind of another swig software ) to generate wrappers for swig generated objects to produce a simple API.

Reply:
Could you please explain the comment in detail..I didn't understand completely

Reply:
> I think most of the work is in writing a script which generates the wrappers automatically

We already have a tool which generates wrappers automatically - it's called "SWIG".

You should be using SWIG to achieve this, not writing extra layers of scripts above or around it.  SWIG is up for the job of automating the majority of the wrapping for all the other languages we use it for - I don't see why Go is different in that regard.